web/application.py

```diff
@@ -111,8 +111,16 @@ def _cleanup(self):
         # Clearing up all thread-local state to avoid interefereing with subsequent requests.
         utils.ThreadedDict.clear_all()
 
+
     def init_mapping(self, mapping):
-        self.mapping = list(utils.group(mapping, 2))
+        if isinstance(mapping, tuple):
+            self.mapping = list(utils.group(mapping, 2))
+        elif isinstance(mapping, dict):
+            self.mapping = list(mapping.iteritems())
+        elif hasattr(mapping, "__iter__") and hasattr(mapping, "append"):
+            self.mapping = mapping
```

Comment:
One INGInious dev found that using a simple tuple for mapping URLs was a bit odd, and probably not the most intuitive data structure.

Here we add two new ways of defining a mapping :
- with a dict (see test_dict_mapping)
- with a custom class, which has to be iterable and support an append(url, view) method (see test_obj_mapping)

The dict option is an easy and intuitive data structure and could become the default way of doing this (if we change the docs). The custom class option is needed by INGInious, where each user has a different mapping, built dynamically according to his role/privileges.